### PR TITLE
feat: add evm env provider

### PR DESCRIPTION
**Pull request by mattsse**
Adds a new provider type that knows how to fill `revm::Env`, this will be required in RPC whenever a transaction is executed (eth_call, eth_estimate etc..)

also adds a new `header_td_by_number` function that queries total difficulty via block number.

builds on #1527

### CODE REVIEW

File: crates/storage/provider/src/traits/evm_env.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+use reth_interfaces::Result;
+use reth_primitives::{BlockId, Header};
+use revm_primitives::{BlockEnv, CfgEnv, Env};
+
+/// A provider type that knows chain specific information required to configure an [Env]
+///
+/// This type is mainly used to provide required data to configure the EVM environment.
+#[auto_impl::auto_impl(&, Arc)]
+pub trait EvmEnvProvider: Send + Sync {
```
Comment (rakita) [inline crates/storage/provider/src/traits/evm_env.rs:9]: This looks niceee!

**Comment by gakonst** [comment]
LGTM modulo conflicts

File: crates/storage/provider/src/providers/mod.rs [hunk 2]
```diff
@@ -231,6 +240,60 @@ impl<DB: Database> WithdrawalsProvider for ShareableDatabase<DB> {
     }
 }
 
+impl<DB: Database> EvmEnvProvider for ShareableDatabase<DB> {
+    fn fill_env_at(&self, env: &mut Env, at: BlockId) -> Result<()> {
+        let hash = self.block_hash_for_id(at)?.ok_or(ProviderError::HeaderNotFound)?;
+        let header = self.header(&hash)?.ok_or(ProviderError::HeaderNotFound)?;
```
Comment (rkrasiuk) [inline crates/storage/provider/src/providers/mod.rs:246]: unrelated to this PR: we should prob accept BlockId in `HeaderProvider::header` and `HeaderProvider::header_td`

**Review by rkrasiuk — APPROVED** [review]
lgtm!

**Comment by rkrasiuk** [comment]
restarted deny check, because it hung up (in progress for ~2h)

**Pull request merged**